File: apps/ios/Pebbles/Features/Path/Services/WeekRollBuilder.swift
import Foundation

/// Pure builder for the iOS Path screen's weeks roll.
///
/// Returns `[WeekRollEntry]` containing every ISO week with at least one
/// pebble, plus the current week and the next week (so the user can always
/// see "today" and "tomorrow's week" in the roll). Entries are sorted
/// ascending by `weekStart`.
///
/// Within each entry, pebbles are sorted asymmetrically:
///   - Past weeks (whose Monday is strictly before the current week's
///     Monday) sort **ascending** by `happenedAt` — reading time forward.
///   - Current and future weeks sort **descending** — most recent first,
///     matching the way the old `PathView` rendered.
enum WeekRollBuilder {

    static func build(
        pebbles: [Pebble],
        calendar: Calendar,
        today: Date
    ) -> [WeekRollEntry] {
        let currentStart = weekStart(for: today, calendar: calendar)
        guard let nextStart = calendar.date(byAdding: .weekOfYear, value: 1, to: currentStart) else {
            return []
        }

        // Bucket pebbles by their week's Monday.
        let grouped: [Date: [Pebble]] = Dictionary(grouping: pebbles) { p in
            weekStart(for: p.happenedAt, calendar: calendar)
        }

        // Union: every week with pebbles, plus current and next.
        let weekStarts = Set(grouped.keys).union([currentStart, nextStart])

        return weekStarts
            .sorted()
            .map { ws in
                let raw = grouped[ws] ?? []
                let sorted = ws < currentStart
                    ? raw.sorted { $0.happenedAt < $1.happenedAt }
                    : raw.sorted { $0.happenedAt > $1.happenedAt }
                return WeekRollEntry(weekStart: ws, pebbles: sorted)
            }
    }

    /// ISO Monday 00:00:00 of the week containing `date`.
    private static func weekStart(for date: Date, calendar: Calendar) -> Date {
        let comps = calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: date)
        var monday = DateComponents()
        monday.yearForWeekOfYear = comps.yearForWeekOfYear
        monday.weekOfYear = comps.weekOfYear
        monday.weekday = 2 // Monday in ISO 8601
        return calendar.date(from: monday) ?? date
    }
}

extension WeekRollBuilder {

    static func previous(of weekStart: Date, in entries: [WeekRollEntry]) -> WeekRollEntry? {
        guard let idx = entries.firstIndex(where: { $0.weekStart == weekStart }), idx > 0 else {
            return nil
        }
        return entries[idx - 1]
    }

    static func next(of weekStart: Date, in entries: [WeekRollEntry]) -> WeekRollEntry? {
        guard let idx = entries.firstIndex(where: { $0.weekStart == weekStart }),
              idx + 1 < entries.count else {
            return nil
        }
        return entries[idx + 1]
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/WeekRollBuilder.kt
package app.pbbls.android.features.path

import app.pbbls.android.features.path.models.Pebble
import app.pbbls.android.features.path.models.WeekRollEntry
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.IsoFields
import java.time.temporal.TemporalAdjusters

/**
 * Pure builder for the Path screen's weeks roll — mirrors iOS
 * `WeekRollBuilder.swift` exactly:
 *
 * Returns every ISO week with at least one pebble, plus the current week (so
 * a fresh user always sees their current-week empty state). Future weeks are
 * NOT pre-created — they only appear once a pebble is dated into them.
 * Entries sort ascending by [WeekRollEntry.weekStart].
 *
 * Within each entry, pebbles sort asymmetrically:
 * - Past weeks (Monday strictly before the current week's Monday) sort
 *   **ascending** by `happenedAt` — reading time forward.
 * - Current and future weeks sort **descending** — most recent first.
 */
object WeekRollBuilder {
    /** ISO Monday of the week containing [date]. */
    fun weekStart(date: LocalDate): LocalDate = date.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))

    /** ISO week-of-week-based-year number of a week's Monday (1..53). */
    fun isoWeekNumber(weekStart: LocalDate): Int = weekStart.get(IsoFields.WEEK_OF_WEEK_BASED_YEAR)

    fun build(
        pebbles: List<Pebble>,
        zone: ZoneId,
        today: LocalDate,
    ): List<WeekRollEntry> {
        val currentStart = weekStart(today)

        // Bucket pebbles by their week's Monday, in the device zone (the
        // iOS Calendar analog — grouping follows the user's wall clock).
        val grouped =
            pebbles.groupBy { pebble ->
                weekStart(pebble.happenedAt.atZoneSameInstant(zone).toLocalDate())
            }

        // Union: every week with pebbles, plus current. No empty future weeks.
        val weekStarts = grouped.keys + currentStart

        return weekStarts
            .sorted()
            .map { ws ->
                val raw = grouped[ws].orEmpty()
                val sorted =
                    if (ws < currentStart) {
                        raw.sortedBy { it.happenedAt.toInstant() }
                    } else {
                        raw.sortedByDescending { it.happenedAt.toInstant() }
                    }
                WeekRollEntry(weekStart = ws, pebbles = sorted)
            }
    }

    fun previous(
        of: LocalDate,
        entries: List<WeekRollEntry>,
    ): WeekRollEntry? {
        val idx = entries.indexOfFirst { it.weekStart == of }
        if (idx <= 0) return null
        return entries[idx - 1]
    }

    fun next(
        of: LocalDate,
        entries: List<WeekRollEntry>,
    ): WeekRollEntry? {
        val idx = entries.indexOfFirst { it.weekStart == of }
        if (idx < 0 || idx + 1 >= entries.size) return null
        return entries[idx + 1]
    }
}
